### PR TITLE
feature: ZENKO-483 Monitor CRR upload

### PR DESCRIPTION
**Pull request by bennettbuchanan**
See feature docs: https://github.com/scality/backbeat/pull/356.

### CODE REVIEW

File: tests/functional/api/BackbeatServer.js [hunk 1]
```diff
@@ -272,6 +272,9 @@ describe('Backbeat Server', () => {
             '/_/metrics/crr/all/completions',
             '/_/metrics/crr/all/failures',
             '/_/metrics/crr/all/throughput',
+            `/_/metrics/crr/${site1}/progress/bucket/object/version`,
```
Comment (bennettbuchanan) [inline tests/functional/api/BackbeatServer.js]: Need to update routing design to allow for a key with a `/` character.

**Comment by bennettbuchanan** [comment]
@bert-e help

**Comment by bennettbuchanan** [comment]
@bert-e wait

File: tests/functional/api/BackbeatServer.js (hunk 1 above)
Comment (bennettbuchanan) [reply tests/functional/api/BackbeatServer.js]: To be updated to use query params instead.

File: extensions/replication/failedCRR/FailedCRRConsumer.js [hunk 2]
```diff
@@ -87,7 +88,57 @@ class FailedCRRConsumer {
             log.end();
             return cb();
         }
-        return this._setRedisKey(data, kafkaEntry, log, cb);
+        return async.series([
+            next => this._removePrexistingKeys(data, err => {
```
Comment (jonathan-gramain) [inline extensions/replication/failedCRR/FailedCRRConsumer.js]: minor typo: missing `e` (`_removePreExistingKeys`)

File: extensions/replication/failedCRR/FailedCRRConsumer.js [hunk 3]
```diff
@@ -87,7 +88,57 @@ class FailedCRRConsumer {
...
+                if (err) {
+                    this.logger.error('error deleting prexisting CRR object ' +
```
Comment (jonathan-gramain) [inline extensions/replication/failedCRR/FailedCRRConsumer.js]: `preexisting`

Maybe a more accurate (and less "scary") formulation could be "error deleting preexisting CRR metrics key"

File: extensions/replication/failedCRR/FailedCRRConsumer.js [hunk 4]
```diff
@@ -87,7 +88,57 @@ class FailedCRRConsumer {
             log.end();
             return cb();
         }
-        return this._setRedisKey(data, kafkaEntry, log, cb);
+        return async.series([
+            next => this._removePrexistingKeys(data, err => {
+                if (err) {
```
Comment (jonathan-gramain) [inline extensions/replication/failedCRR/FailedCRRConsumer.js]: Not sure, if the key does not exist does it return an error? In which case it should be filtered to avoid logging it (as I think this is a normal case).

File: extensions/replication/queueProcessor/task.js [hunk 5]
```diff
@@ -157,14 +158,13 @@ function initAndStart(zkClient) {
...
-                            repConfig, redisConfig, site);
+                            repConfig, redisConfig, mConfig, site);
                         activeQProcessors[site] = qp;
                         setupZkSiteNode(qp, zkClient, site, (err, data) => {
                             if (err) {
                                 return next(err);
                             }
-                            qp.start({ paused: data.paused });
-                            return next();
+                            return qp.start({ paused: data.paused }, next);
```
Comment (jonathan-gramain) [inline extensions/replication/queueProcessor/task.js]: Why this change since `QueuePopulator.start()` does not appear to take a callback argument

File: extensions/replication/tasks/ReplicateObject.js [hunk 6]
```diff
@@ -359,7 +362,23 @@ class ReplicateObject extends BackbeatTask {
                 return doneOnce(err);
             }
             partObj.setDataLocation(data.Location[0]);
-            return doneOnce(null, partObj.getValue());
+            return async.series([
+                next => {
+                    if (partNumber === 1) {
+                        const extMetrics = getExtMetrics(this.site,
```
Comment (jonathan-gramain) [inline extensions/replication/tasks/ReplicateObject.js]: Maybe you should consider doing this at the beginning of the process rather than after having put the first part (if possible), so you have the total available as soon as the object starts being put, instead of after the first part is complete. Note that here you can also have some completion metrics written for other parts even before the "queued" key is written, e.g. if part #2 and #3 are completed but not #1 yet. Not sure this has an important impact but just want to point it out, and pushing the "queued" key first would avoid this also.

File: lib/api/BackbeatAPI.js [hunk 7]
```diff
@@ -157,13 +162,29 @@ class BackbeatAPI {
...
+        }
 
         // All routes at this point have extensions property in use
         // Match http verb and match extension name
-        let filteredRoutes = routes.filter(r => (
-            bbRequest.getHTTPMethod() === r.httpMethod &&
-            r.extensions[rDetails.extension]
-        ));
+        let filteredRoutes = routes.filter(r => {
+            const isCorrectMethod = bbRequest.getHTTPMethod() === r.httpMethod;
```
Comment (jonathan-gramain) [inline lib/api/BackbeatAPI.js]: minor naming issue: "correct" may not be the appropriate wording, since we are trying to find a route that "matches" what we provide, it's not really "incorrect" not to match a specific route. Maybe `isMatchingXxx` instead would work better.

**Review by jonathan-gramain — CHANGES_REQUESTED** [review]

File: extensions/replication/failedCRR/FailedCRRConsumer.js (hunk 4 above)
Comment (bennettbuchanan) [reply extensions/replication/failedCRR/FailedCRRConsumer.js]: If the key doesn't exist there is no error, just an empty response. The error would occur if the fetch from Redis failed for some reason (e.g. Redis is offline).

File: extensions/replication/queueProcessor/task.js (hunk 5 above)
Comment (bennettbuchanan) [reply extensions/replication/queueProcessor/task.js]: Nice catch, this is left-over from a rebase. I've removed it.

File: extensions/replication/tasks/ReplicateObject.js (hunk 6 above)
Comment (bennettbuchanan) [reply extensions/replication/tasks/ReplicateObject.js]: Very good point. I've updated this.

File: tests/functional/api/BackbeatServer.js [hunk 8]
```diff
@@ -542,6 +551,30 @@ describe('Backbeat Server', () => {
...
         });
 
+        it(`should return all metrics for route: /_/metrics/crr/${site1}` +
+            `/progress/${BUCKET_NAME}/${OBJECT_KEY}?versionId=${VERSION_ID}`,
+            done =>
+            getRequest(`/_/metrics/crr/${site1}/progress/${BUCKET_NAME}/` +
+                `${OBJECT_KEY}?versionId=${VERSION_ID}`, (err, res) => {
+                assert.ifError(err);
+                assert(res.description);
+                assert.equal(res.pending, 50);
```
Comment (jonathan-gramain) [inline tests/functional/api/BackbeatServer.js]: minor: we usually use `strictEqual` to also check the type

**Review by jonathan-gramain — APPROVED** [review]

File: extensions/replication/failedCRR/FailedCRRConsumer.js [hunk 9]
```diff
@@ -87,7 +88,57 @@ class FailedCRRConsumer {
...
+     * @return {undefined}
+     */
+    _removePreexistingKeys(data, cb) {
+        // eslint-disable-next-line no-unused-vars
+        const [service, extension, status, bucketName, objectKey, versionId,
+            site] = data.key.split(':');
+        // CRR object-level metrics key schema.
+        const queryString = `${site}:${bucketName}:${objectKey}:${versionId}:` +
+            `${service}:${extension}:*`;
+        return redisClient.batch([['keys', queryString]], (err, res) => {
```
Comment (rahulreddy) [inline extensions/replication/failedCRR/FailedCRRConsumer.js]: Using `keys *` pattern is a bad idea to use on a production Redis. Please use `scan` instead.
Reference from redis.io/commands/keys

> Warning: consider KEYS as a command that should only be used in production environments with extreme care. It may ruin performance when it is executed against large databases.

**Review by rahulreddy — CHANGES_REQUESTED** [review]

File: extensions/replication/failedCRR/FailedCRRConsumer.js (hunk 9 above)
Comment (bennettbuchanan) [reply extensions/replication/failedCRR/FailedCRRConsumer.js]: Good call, scan stream is handy.

**Review by rahulreddy — APPROVED** [review]
